### PR TITLE
Fix/fugacity

### LINKED ISSUE #1257: Make fugacity a single type parameter.
## Description

In HPMC, change all instances of `fugacity[i,j]` to `fugacity[i]`

## Motivation and context

See the explanation in #659.

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Switches fugacity in HPMC integrators to be per-type instead of per-type-pair.

## Motivation and context
Per-type-pair depletion interactions break the n-body depletion potential, see #659.

Resolves #1257

## Change log

```
Changed: [breaking] Change fugacity to a per-type quantity
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/UpdaterClustersGPU.h [hunk 1]
```diff
@@ -354,20 +349,13 @@ template<class Shape> void UpdaterClustersGPU<Shape>::update(uint64_t timestep)
...
-                max_d = std::max(
-                    max_d,
-                    (Scalar)0.5
-                        * (tmp_i.getCircumsphereDiameter() + tmp_j.getCircumsphereDiameter()));
-                }
+            continue;
             }
+        quat<Scalar> o;
+        Shape tmp_i(o, params[type_i]);
+        max_d = std::max(max_d, (Scalar)tmp_i.getCircumsphereDiameter());
```
Comment (b-butler) [inline hoomd/hpmc/UpdaterClustersGPU.h]: static cast

File: hoomd/hpmc/IntegratorHPMCMono.h [hunk 2]
```diff
@@ -1635,17 +1613,12 @@ void IntegratorHPMCMono<Shape>::updateCellWidth()
...
-            Shape tmp_i(o, this->m_params[type_i]);
-            if (m_fugacity[m_depletant_idx(type_i,type_j)] != Scalar(0.0))
-                {
-                // add range of depletion interaction
-                Shape tmp_j(o, this->m_params[type_j]);
-                max_d = std::max(max_d, (Scalar) 0.5*(tmp_i.getCircumsphereDiameter() +
-                    tmp_j.getCircumsphereDiameter()));
-                }
+            Shape tmp_shape(o, this->m_params[type_i]);
+            max_d = std::max(max_d, (Scalar)tmp_shape.getCircumsphereDiameter()); // maybe need (Scalar) here
```
Comment (b-butler) [inline hoomd/hpmc/IntegratorHPMCMono.h]: static cast

File: hoomd/hpmc/IntegratorHPMCMono.h [hunk 3]
```diff
@@ -2011,12 +1984,14 @@ inline bool IntegratorHPMCMono<Shape>::checkDepletantOverlap(unsigned int i, vec
...
             // GlobalVector is not thread-safe, access it outside the parallel loop
-            Scalar fugacity = m_fugacity[m_depletant_idx(type_a,type_b)];
+            //Scalar fugacity = m_fugacity[m_depletant_idx(type_a,type_b)];
+            Scalar fugacity = m_fugacity[type_a];
             if (fugacity == 0.0 || (!h_overlaps[this->m_overlap_idx(type_a, typ_i)]
                 && !h_overlaps[this->m_overlap_idx(type_b, typ_i)]))
                 continue;
 
-            unsigned int ntrial = m_ntrial[m_depletant_idx(type_a,type_b)];
+            //unsigned int ntrial = m_ntrial[m_depletant_idx(type_a,type_b)];
```
Comment (b-butler) [inline hoomd/hpmc/IntegratorHPMCMono.h]: remove commented lines.

File: hoomd/hpmc/IntegratorHPMCMono.h [hunk 4]
```diff
@@ -2306,9 +2281,11 @@ inline bool IntegratorHPMCMono<Shape>::checkDepletantOverlap(unsigned int i, vec
                         if (! shape_i.ignoreStatistics())
                             {
                             #ifdef ENABLE_TBB
-                            thread_implicit_counters[m_depletant_idx(type_a,type_b)].local().insert_count++;
+                            //thread_implicit_counters[m_depletant_idx(type_a,type_b)].local().insert_count++;
```
Comment (b-butler) [inline hoomd/hpmc/IntegratorHPMCMono.h]: remove commented lines

File: hoomd/hpmc/IntegratorHPMCMonoGPU.h [hunk 5]
```diff
@@ -1269,14 +1244,13 @@ template<class Shape> void IntegratorHPMCMonoGPU<Shape>::update(uint64_t timeste
                     this->m_exec_conf->beginMultiGPU();
                     for (unsigned int itype = 0; itype < this->m_pdata->getNTypes(); ++itype)
                         {
+                        if (h_fugacity.data[itype] == 0)
+                            continue;
                         for (unsigned int jtype = itype; jtype < this->m_pdata->getNTypes();
                              ++jtype)
```
Comment (b-butler) [inline hoomd/hpmc/IntegratorHPMCMonoGPU.h]: Check for need of this loop.

**Review by b-butler — COMMENTED** [review]
Initial look through after taking over for Tim.

**Comment by b-butler** [comment]
@tcmoore3 I have refactored the necessary code, I believe. I think we wanted a validation test as well. I am not sure what result to validate off of though as I have never directly dealt with depletants. Did you have an idea?

**Comment by joaander** [comment]
v2 had a validation test. We could port those numbers over to the new separate hoomd-validation repository.

File: hoomd/hpmc/IntegratorHPMCMonoGPUDepletantsAuxilliaryPhase2.cuh [hunk 6]
```diff
@@ -537,36 +505,6 @@ __launch_bounds__(max_threads)
                 bool overlap_i_a = true;
                 bool overlap_i_b = true;
 
-                if (pairwise)
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMCMonoGPUDepletantsAuxilliaryPhase2.cuh:540]: Remove the unused `pairwise` template parameter.

File: hoomd/hpmc/IntegratorHPMCMonoGPUDepletantsAuxilliaryPhase1.cuh [hunk 7]
```diff
@@ -550,40 +522,6 @@ __launch_bounds__(max_threads)
                 bool overlap_i_a = true;
                 bool overlap_i_b = true;
 
-                if (pairwise)
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMCMonoGPUDepletantsAuxilliaryPhase1.cuh:553]: Remove the unused `pairwise` template parameter.

File: hoomd/hpmc/IntegratorHPMCMonoGPUDepletants.cuh [hunk 8]
```diff
@@ -281,37 +270,15 @@ __launch_bounds__(max_threads)
                                   && check_circumsphere_overlap(r_ij, shape_test_a, shape_i)
                                   && test_overlap(r_ij, shape_test_a, shape_i, err_count));
 
-            bool overlap_old_b = overlap_old_a;
-            if (pairwise)
```
Comment (joaander) [inline hoomd/hpmc/IntegratorHPMCMonoGPUDepletants.cuh:285]: Remove the unused `pairwise` template parameter.

**Review by joaander — COMMENTED** [review]
Thanks @b-butler and @tcmoore3 ! I have a few minor comments.

@jglaser please review this and let us know if we made the correct changes.

**Review by joaander — APPROVED** [review]
Thanks!

@jglaser we would appreciate a review to ensure we made these changes correctly. I will merge this on Wednesday, April 13 with or without your review.

**Comment by jglaser** [comment]
Hi, I've compiled his branch on Summit. However, it seems like the validation test has not been updated. Without a proper validation test the depletants functionality is useless. It may alternatively be removed, if the maintenance burden is too high.

Here is the output I get from the test in `hpmc/pytest/depletants_sphere.py`

```
(/gpfs/alpine/world-shared/bip214/hoomd) [glaser@login1.summit hoomd-blue]$ LD_PRELOAD=${OLCF_GCC_ROOT}/lib64/libstdc++.so.6 python -m pytest build/hoomd/hpmc/pytest/depletants_sphere.py 
================================================= test session starts ==================================================
platform linux -- Python 3.7.13, pytest-7.1.1, pluggy-1.0.0
rootdir: /gpfs/alpine/bip214/proj-shared/hoomd-blue
collected 4 items                                                                                                      

build/hoomd/hpmc/pytest/depletants_sphere.py FFEE                                                                [100%]

======================================================== ERRORS ========================================================
__________________________________ ERROR at setup of test_sphere[GPU-False-1-0.4-0.1] __________________________________

request = <SubRequest 'device' for <Function test_sphere[GPU-False-1-0.4-0.1]>>

    @pytest.fixture(scope='session', params=devices)
    def device(request):
        """Parameterized Device fixture.
    
        Tests that use `device` will run once on the CPU and once on the GPU. The
        device object is session scoped to avoid device creation overhead when
        running tests.
        """
>       d = request.param()

build/hoomd/conftest.py:45: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hoomd.device.GPU object at 0x7ff713513810>, gpu_ids = [], num_cpu_threads = None, communicator = None
msg_file = None, notice_level = 2

    def __init__(self,
                 gpu_ids=None,
                 num_cpu_threads=None,
                 communicator=None,
                 msg_file=None,
                 notice_level=2):
    
        super().__init__(communicator, notice_level, msg_file)
    
        if gpu_ids is None:
            gpu_ids = []
    
        # convert None options to defaults
        self._cpp_exec_conf = _hoomd.ExecutionConfiguration(
            _hoomd.ExecutionConfiguration.executionMode.GPU, gpu_ids,
>           self.communicator.cpp_mpi_conf, self._cpp_msg)
E       RuntimeError: HIP Error

build/hoomd/device.py:232: RuntimeError
------------------------------------------------ Captured stderr setup -------------------------------------------------
**ERROR**: unknown error before /hoomd/ExecutionConfiguration.cc:351
__________________________________ ERROR at setup of test_sphere[GPU-True-1-0.4-0.1] ___________________________________

request = <SubRequest 'device' for <Function test_sphere[GPU-False-1-0.4-0.1]>>

    @pytest.fixture(scope='session', params=devices)
    def device(request):
        """Parameterized Device fixture.
    
        Tests that use `device` will run once on the CPU and once on the GPU. The
        device object is session scoped to avoid device creation overhead when
        running tests.
        """
>       d = request.param()

build/hoomd/conftest.py:45: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hoomd.device.GPU object at 0x7ff713513810>, gpu_ids = [], num_cpu_threads = None, communicator = None
msg_file = None, notice_level = 2

    def __init__(self,
                 gpu_ids=None,
                 num_cpu_threads=None,
                 communicator=None,
                 msg_file=None,
                 notice_level=2):
    
        super().__init__(communicator, notice_level, msg_file)
    
        if gpu_ids is None:
            gpu_ids = []
    
        # convert None options to defaults
        self._cpp_exec_conf = _hoomd.ExecutionConfiguration(
            _hoomd.ExecutionConfiguration.executionMode.GPU, gpu_ids,
>           self.communicator.cpp_mpi_conf, self._cpp_msg)
E       RuntimeError: HIP Error

build/hoomd/device.py:232: RuntimeError
======================================================= FAILURES =======================================================
___________________________________________ test_sphere[CPU-False-1-0.4-0.1] ___________________________________________

simulation_factory = <function simulation_factory.<locals>.make_simulation at 0x7ff70b7307a0>
lattice_snapshot_factory = <function lattice_snapshot_factory.<locals>.make_snapshot at 0x7ff70b7308c0>, phi = 0.1
eta_p = 0.4, ntrial = 1, use_clusters = False

    @pytest.mark.parametrize("phi", [0.1])
    @pytest.mark.parametrize("eta_p", [0.4])
    @pytest.mark.parametrize("ntrial", [1])
    @pytest.mark.parametrize("use_clusters", [False, True])
    def test_sphere(simulation_factory, lattice_snapshot_factory, phi, eta_p,
                    ntrial, use_clusters):
        # number of spheres
        n = 5
        N = n**3
        d_sphere = 1.0
        V_sphere = math.pi / 6.0 * math.pow(d_sphere, 3.0)
    
        # depletant-colloid size ratio (unity, to emphasize many-body effects)
        q = 1.0
    
        L_target = math.pow(N * V_sphere / phi, 1.0 / 3.0)
    
        a = L_target / n
        sim = simulation_factory(
            lattice_snapshot_factory(particle_types=['A', 'B'],
                                     dimensions=3,
                                     a=a,
                                     n=n))
    
>       mc = hoomd.hpmc.integrate.Sphere(seed=1, default_d=0.1, default_a=0.1)
E       TypeError: __init__() got an unexpected keyword argument 'seed'

build/hoomd/hpmc/pytest/depletants_sphere.py:98: TypeError
___________________________________________ test_sphere[CPU-True-1-0.4-0.1] ____________________________________________

simulation_factory = <function simulation_factory.<locals>.make_simulation at 0x7ff70b7307a0>
lattice_snapshot_factory = <function lattice_snapshot_factory.<locals>.make_snapshot at 0x7ff70b7308c0>, phi = 0.1
eta_p = 0.4, ntrial = 1, use_clusters = True

    @pytest.mark.parametrize("phi", [0.1])
    @pytest.mark.parametrize("eta_p", [0.4])
    @pytest.mark.parametrize("ntrial", [1])
    @pytest.mark.parametrize("use_clusters", [False, True])
    def test_sphere(simulation_factory, lattice_snapshot_factory, phi, eta_p,
                    ntrial, use_clusters):
        # number of spheres
        n = 5
        N = n**3
        d_sphere = 1.0
        V_sphere = math.pi / 6.0 * math.pow(d_sphere, 3.0)
    
        # depletant-colloid size ratio (unity, to emphasize many-body effects)
        q = 1.0
    
        L_target = math.pow(N * V_sphere / phi, 1.0 / 3.0)
    
        a = L_target / n
        sim = simulation_factory(
            lattice_snapshot_factory(particle_types=['A', 'B'],
                                     dimensions=3,
                                     a=a,
                                     n=n))
    
>       mc = hoomd.hpmc.integrate.Sphere(seed=1, default_d=0.1, default_a=0.1)
E       TypeError: __init__() got an unexpected keyword argument 'seed'

build/hoomd/hpmc/pytest/depletants_sphere.py:98: TypeError
=============================================== short test summary info ================================================
FAILED build/hoomd/hpmc/pytest/depletants_sphere.py::test_sphere[CPU-False-1-0.4-0.1] - TypeError: __init__() got an ...
FAILED build/hoomd/hpmc/pytest/depletants_sphere.py::test_sphere[CPU-True-1-0.4-0.1] - TypeError: __init__() got an u...
ERROR build/hoomd/hpmc/pytest/depletants_sphere.py::test_sphere[GPU-False-1-0.4-0.1] - RuntimeError: HIP Error
ERROR build/hoomd/hpmc/pytest/depletants_sphere.py::test_sphere[GPU-True-1-0.4-0.1] - RuntimeError: HIP Error
============================================= 2 failed, 2 errors in 0.30s ==============================================
```

**Comment by joaander** [comment]
> Hi, I've compiled his branch on Summit. However, it seems like the validation test has not been updated. Without a proper validation test the depletants functionality is useless. It may alternatively be removed, if the maintenance burden is too high.
> 
> Here is the output I get from the test in `hpmc/pytest/depletants_sphere.py`

`depletants_sphere.py` is not present in the source branch.

Validation tests that run within the time constraints of GitHub actions platforms are worse than useless. They cannot possibly run long enough to sample equilibrium and therefore provide both false positive and false negative results. We have plans to port all validation tests, make them correct, and add new ones to a separate repository that runs for sufficient length on HPC platforms: #210 However, we do not have infinite amounts of developer effort available so we have not completed this work yet. We will complete it in due time and include validation tests for the sphere depletants.

Do the modifications to the code look correct?

**Comment by joaander** [comment]
I added a note to bring the spherical depletants test forward to hoomd-validation in #210.

**Pull request merged**